File: src/services/quoteService.js
const urgencyMultiplier = {
  normal: { price: 1, days: 0 },
  expressa: { price: 1.25, days: -1 },
  urgente: { price: 1.45, days: -2 },
}

const deliveryFees = {
  retirada: 0,
  entrega: 28,
}

export function calculateQuote(product, config) {
  if (!product) {
    return {
      subtotal: 0,
      urgencyFee: 0,
      deliveryFee: 0,
      total: 0,
      estimatedDays: 0,
      details: [],
    }
  }

  const quantity = Number(config.quantity || product.quantities[0] || 1)
  const quantityFactor = Math.max(quantity / 100, 1)
  const size = product.sizes.find((item) => item.name === config.size) ?? product.sizes[0]
  const material = product.materials.find((item) => item.name === config.material) ?? product.materials[0]
  const printMode = product.printModes.find((item) => item.name === config.printMode) ?? product.printModes[0]
  const finishing = product.finishings.find((item) => item.name === config.finishing) ?? product.finishings[0]
  const urgency = urgencyMultiplier[config.urgency] ?? urgencyMultiplier.normal
  const deliveryFee = deliveryFees[config.delivery] ?? 0

  const base = product.basePrice * quantityFactor
  const variable = size.price + material.price + printMode.price + finishing.price
  const subtotal = Math.round((base + variable * quantityFactor) * 100) / 100
  const urgencyFee = Math.round((subtotal * (urgency.price - 1)) * 100) / 100
  const total = Math.round((subtotal + urgencyFee + deliveryFee) * 100) / 100
  const estimatedDays = Math.max(1, product.baseDeadline + size.days + material.days + finishing.days + urgency.days)

  return {
    subtotal,
    urgencyFee,
    deliveryFee,
    total,
    estimatedDays,
    details: [
      `${quantity.toLocaleString('pt-BR')} unidades`,
      size.name,
      material.name,
      printMode.name,
      finishing.name,
      config.delivery === 'entrega' ? 'Entrega local' : 'Retirada no balcão',
    ],
  }
}

File: API/PrintFlowApi/Services/QuoteService.cs
using PrintFlowApi.DTOs;
using PrintFlowApi.Model;

namespace PrintFlowApi.Services;

public class QuoteService(IConfiguration configuration)
{
    private static readonly Dictionary<string, (decimal Multiplier, int Days)> Urgency = new(StringComparer.OrdinalIgnoreCase)
    {
        ["normal"] = (1m, 0),
        ["expressa"] = (1.25m, -1),
        ["urgente"] = (1.45m, -2)
    };

    public QuoteResponse Calculate(Product product, QuoteRequest request)
    {
        var size = FindOption(product, "size", request.Size);
        var material = FindOption(product, "material", request.Material);
        var printMode = FindOption(product, "printMode", request.PrintMode);
        var finishing = FindOption(product, "finishing", request.Finishing);
        var urgency = Urgency.GetValueOrDefault(request.Urgency, Urgency["normal"]);
        var deliveryFee = request.Delivery == DeliveryMode.LocalDelivery
            ? configuration.GetValue("Business:DeliveryFee", 28m)
            : 0m;

        var quantityFactor = Math.Max(request.Quantity / 100m, 1m);
        var basePrice = product.BasePrice * quantityFactor;
        var variable = size.PriceDelta + material.PriceDelta + printMode.PriceDelta + finishing.PriceDelta;
        var subtotal = Round(basePrice + variable * quantityFactor);
        var urgencyFee = Round(subtotal * (urgency.Multiplier - 1m));
        var total = Round(subtotal + urgencyFee + deliveryFee);
        var estimatedDays = Math.Max(1, product.BaseDeadlineDays + size.DeadlineDeltaDays + material.DeadlineDeltaDays + finishing.DeadlineDeltaDays + urgency.Days);

        return new QuoteResponse(
            subtotal,
            urgencyFee,
            deliveryFee,
            total,
            estimatedDays,
            [
                $"{request.Quantity:N0} unidades",
                size.Name,
                material.Name,
                printMode.Name,
                finishing.Name,
                request.Delivery == DeliveryMode.LocalDelivery ? "Entrega local" : "Retirada no balcao"
            ]);
    }

    private static ProductOption FindOption(Product product, string type, string name)
    {
        return product.Options.FirstOrDefault(option =>
                option.Type.Equals(type, StringComparison.OrdinalIgnoreCase) &&
                option.Name.Equals(name, StringComparison.OrdinalIgnoreCase))
            ?? throw new InvalidOperationException($"Opcao invalida para {type}: {name}");
    }

    private static decimal Round(decimal value) => Math.Round(value, 2, MidpointRounding.AwayFromZero);
}
